google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutations.java

```diff
@@ -0,0 +1,137 @@
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.models;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.InternalApi;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.common.base.Preconditions;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import org.threeten.bp.Duration;
+
+/**
+ * Tracker for outstanding bulk mutations. Allows for the caller to wait for all of the outstanding
+ * mutations to complete.
+ *
+ * @see com.google.cloud.bigtable.data.v2.BigtableDataClient#newBulkMutations() for example usage.
+ */
+public final class BulkMutations implements AutoCloseable {
+  private final UnaryCallable<RowMutation, Void> callable;
+  // Shared
+  private final Object lock = new Object();
+  private final AtomicLong numOutstanding = new AtomicLong();
+  // Only accessed by grpc
+  private long numFailures;
+  // Only accessed by caller
+  private boolean closed;
+
+  @InternalApi
+  public BulkMutations(UnaryCallable<RowMutation, Void> callable) {
+    this.callable = callable;
+  }
+
+  /**
+   * Waits up to 10 minutes for all of the outstanding mutations to be resolved.
+   *
+   * @see #close(Duration) for more infortmation.
+   */
+  @Override
+  public void close() throws InterruptedException, TimeoutException {
+    close(Duration.ofMinutes(10));
+  }
+
+  /**
+   * Prevents further mutations and waits for all outstanding mutations to complete.
+   *
+   * @throws BulkMutationFailure If any mutations failed.
+   * @throws InterruptedException If interrupted.
+   * @throws TimeoutException If the outstanding requests don't finish in time.
+   */
+  public void close(Duration duration) throws InterruptedException, TimeoutException {
+    closed = true;
+    long deadlineMs = System.currentTimeMillis() + duration.toMillis();
+
+    synchronized (lock) {
+      while (numOutstanding.get() > 0) {
+        long waitMs = deadlineMs - System.currentTimeMillis();
+        if (waitMs <= 0) {
+          throw new TimeoutException("Timed out waiting outstanding mutations to finish");
+        }
+        lock.wait(waitMs);
+      }
+
+      if (numFailures > 0) {
+        throw new BulkMutationFailure(numFailures);
+      }
+    }
+  }
+
+  /**
+   * Queues the mutation to be batched and sent. Please note that the order of mutations is not
+   * guaranteed.
+   *
+   * @throws IllegalStateException If this instance has been closed.
+   */
+  public ApiFuture<Void> send(final RowMutation rowMutation) {
+    Preconditions.checkState(!closed, "BulkMutations has been closed");
+    numOutstanding.incrementAndGet();
+
+    ApiFuture<Void> future = callable.futureCall(rowMutation);
+
+    ApiFutures.addCallback(
+        future,
+        new ApiFutureCallback<Void>() {
+          @Override
+          public void onFailure(Throwable throwable) {
```

Comment:
I have commented something similar before. Don't we want to capture some errors here?
All users will see right now is number of failures and have no idea what the errors are.
If you are planning to add more detailed logging later, perhaps adding a TODO?

Reply:
I tried to address this in the javadoc for `close()`. Not all usecases need to track per item exceptions. For example beam only cares if a bundle failed as a whole. And it requires possibly unbounded resources. For example if the application is a long running stream that blindly writes to bigtable, then this class will leak memory by accumulating errors that never get checked.
Instead the user should opt-in to tracking the individual errors by tracking the futures returned from send. The coarse grained tracking done here is meant to use constant memory and is simply a safety net to remind the callers that there might data loss